In [1]:
import tabulate
from pathlib import Path
import numpy as np

from typing import List, Dict
from collections import defaultdict

import tabulate

In [2]:
models = ['mlp_wass', 'mlp_js', 'mlp_ring']

step_sizes = [0.0001, 0.0001, 0.001]

features = {
    "dumb": [""],
    "discriminator": [""],
    "cluster": {
         "": ["",],# "_dis"],
        "_v1": ["", "_dis"],
        "_v2": {
            "": ["", "_dis"],
            # "_lin": ["", "_dis"],
            # "_quad": ["", "_dis"],
        },
        # "_v3": ["", "_dis"]
    },
    "mmd": {
            "": ["", "_dis"],
            # "_lin": [""], #"_dis"],
            # "_quad": [""],#, "_dis"],
        },
    "cmd": {
            "": ["", "_dis"],
        }, 
}
target = "Discriminator"

In [3]:
log_dir = '../log'

In [4]:
exps_mean = defaultdict(list)
exps_std = defaultdict(list)

for model, step_size in zip(models, step_sizes):
    exps_mean['models'].append(model)
    exps_std['models'].append(model)

    path = Path(log_dir, f'dumb_feature_{target}Target', f'{model}_{step_size}')
    try:
        emd = np.loadtxt(Path(path, 'EMD.txt'))[:, 0]
        std = emd.std(ddof=1)
        emd = emd.mean()
    except:
        emd = np.loadtxt(Path(path, 'callback_results.txt'))[2, 0]
    exps_mean['gan'].append(emd)
    exps_std['gan'].append(std)

def move(subs, suff, feature):
    if isinstance(subs, List):
        for model, step_size in zip(models, step_sizes):
            for s in subs:
                path = Path(log_dir, f'{feature}_feature_{target}Target', f'{model}{suff}{s}_{step_size}')
                #try:
                emd = np.loadtxt(Path(path, 'EMD.txt'))[:, 1]
                std = emd.std(ddof=1)
                emd = emd.mean()
                # except:
                #     emd = np.loadtxt(Path(path, 'callback_results.txt'))[2, -1]
                exps_mean[feature+path.name[len(model):-len(f'_{step_size}')]].append(emd)
                exps_std[feature+path.name[len(model):-len(f'_{step_size}')]].append(std)
        
            # exps.extend([Path(log_dir, f'{feature}_feature_{target}Target', f'{model}{suff}{s}_0.001') for s in subs])
        return
    elif isinstance(subs, Dict):
        for sub_key, subs_ in subs.items():
            move(subs_, suff+sub_key, feature)

for feature, subs in features.items():
    move(subs, "", feature)


In [5]:

exps_mean_cp = {}
for k, v in exps_mean.items():
    k = k.replace('discriminator', 'dis')
    k = k.replace('dis', 'dis')
    k = k.replace('cluster', 'clust')
    exps_mean_cp[k] = v

exps_std_cp = {}
for k, v in exps_std.items():
    k = k.replace('discriminator', 'dis')
    k = k.replace('dis', 'dis')
    k = k.replace('cluster', 'clust')
    exps_std_cp[k] = v

In [6]:
# tabulate.tabulate(exps_mean_cp, tablefmt='latex', headers='keys', floatfmt='.2')

In [7]:
# print(tabulate.tabulate(exps_cp, tablefmt='grid', headers='keys', floatfmt='.2'))

In [8]:
print(tabulate.tabulate(exps_mean_cp, tablefmt='pipe', headers='keys', floatfmt='.3'))

| models   |   gan |   dumb |   dis |   clust |   clust_v1 |   clust_v1_dis |   clust_v2 |   clust_v2_dis |   mmd |   mmd_dis |   cmd |   cmd_dis |
|:---------|------:|-------:|------:|--------:|-----------:|---------------:|-----------:|---------------:|------:|----------:|------:|----------:|
| mlp_wass | 0.26  |  0.251 | 0.296 |   0.233 |      0.243 |          0.237 |      0.254 |          0.23  | 0.276 |     0.252 | 0.246 |     0.215 |
| mlp_js   | 0.373 |  0.433 | 0.437 |   0.311 |      0.352 |          0.355 |      0.314 |          0.301 | 0.41  |     0.36  | 0.306 |     0.297 |
| mlp_ring | 2.11  |  2.2   | 1.52  |   2.2   |      1.82  |          1.49  |      2.03  |          1.94  | 2.22  |     2.22  | 2.23  |     2.15  |


In [9]:
print(tabulate.tabulate(exps_std_cp, tablefmt='pipe', headers='keys', floatfmt='.3'))

| models   |    gan |    dumb |    dis |   clust |   clust_v1 |   clust_v1_dis |   clust_v2 |   clust_v2_dis |    mmd |   mmd_dis |     cmd |   cmd_dis |
|:---------|-------:|--------:|-------:|--------:|-----------:|---------------:|-----------:|---------------:|-------:|----------:|--------:|----------:|
| mlp_wass | 0.0172 | 0.0189  | 0.0466 | 0.00929 |     0.017  |        0.00795 |     0.0123 |         0.0079 | 0.0202 |    0.0219 | 0.00947 |   0.00271 |
| mlp_js   | 0.0357 | 0.0584  | 0.0226 | 0.019   |     0.0101 |        0.00448 |     0.0142 |         0.015  | 0.0306 |    0.0421 | 0.0222  |   0.0136  |
| mlp_ring | 0.0384 | 0.00888 | 0.046  | 0.0477  |     0.0364 |        0.00261 |     0.0324 |         0.0343 | 0.032  |    0.016  | 0.0177  |   0.0286  |


| models   |   gan |   dumb |   dis |   clust |   clust_v1 |   clust_v1_dis |   clust_v2 |   clust_v2_dis |   mmd |   mmd_dis |   cmd |   cmd_dis |
|:---------|------:|-------:|------:|--------:|-----------:|---------------:|-----------:|---------------:|------:|----------:|------:|----------:|
| mlp_wass | 0.26  |  0.251 | 0.296 |   0.233 |      0.243 |          0.237 |      0.254 |          0.23  | 0.276 |     0.252 | 0.246 |     **0.215** |
| mlp_js   | 0.373 |  0.433 | 0.437 |   0.311 |      0.352 |          0.355 |      0.314 |          **0.301** | 0.41  |     0.36  | **0.306** |     **0.297** |
| mlp_ring | 2.11  |  2.2   | 1.52  |   2.2   |      1.82  |          **1.49**  |      2.03  |          1.94  | 2.22  |     2.22  | 2.23  |     2.15  |

| models   |   gan |   dumb |   dis |   clust |   clust_dis |   clust_v1 |   clust_v1_dis |   clust_v2 |   clust_v2_dis |   mmd |   mmd_dis |   cmd |   cmd_dis |
|:---------|------:|-------:|------:|--------:|------------:|-----------:|---------------:|-----------:|---------------:|------:|----------:|------:|----------:|
| mlp_wass | 0.254 |  0.253 | 0.318 |   0.217 |       0.251 |      0.247 |          0.238 |      0.239 |          0.221 | 0.268 |     0.258 | 0.22  |     0.231 |
| mlp_js   | 0.366 |  0.44  | 0.471 |   0.313 |       0.384 |      0.369 |          0.364 |      0.316 |          0.343 | 0.513 |     0.39  | 0.323 |     0.296 |
| mlp_ring | 2.81  |  2.81  | 2.8   |   2.81  |       2.77  |      2.03  |          1.47  |      2.37  |          2.14  | 2.81  |     2.8   | 2.81  |     2.7   |


| models   |   gan |   dumb |   dis |   clust |   clust_dis |   clust_v1 |   clust_v1_dis |   clust_v2 |   clust_v2_dis |   mmd |   mmd_dis |   cmd |   cmd_dis |
|:---------|------:|-------:|------:|--------:|------------:|-----------:|---------------:|-----------:|---------------:|------:|----------:|------:|----------:|
| mlp_wass |  0.25 |   0.25 |  0.32 |   <font color='red'>0.22</font> |        0.25 |       0.25 |           0.24 |       0.24 |           <font color='red'>0.22</font> |  0.27 |      0.26 |  <font color='red'>0.22</font> |      0.23 |
| mlp_js   |  0.37 |   0.44 |  0.47 |    <font color='red'>0.31</font> |        0.38 |       0.37 |           0.36 |       0.32 |           0.34 |  0.51 |      0.39 |  0.32 |      <font color='red'>0.3</font>  |
| mlp_ring |  2.8  |   2.8  |  2.8  |    2.8  |        2.8  |       2.0  |           <font color='red'>1.5</font>  |       2.4  |           2.2  |  2.8  |      2.8  |  2.8  |      2.4  |

clust:

$F_k(x) = \sigma(\|x - \mu_k\|_2 - \Sigma_k),~k=1..K$

clust v1:

$F_k(x) = \|x - \mu_k\|^2_2, ~k=1..K$

clust v2:

$F_k(x) = k(x, x_I) - \frac{2}{K} k(x, \mu_k) $

| models   |   dumb |   cl |   cl_dis |   cl_v1 |   cl_v1_dis |   cl_v2 |   cl_v2_dis |   cl_v2_lin |   cl_v2_quad |   mmd |   mmd_dis |   mmd_lin |   mmd_lin_dis |   mmd_quad |   cmd |   cmd_dis |
|:---------|-------:|-----:|---------:|--------:|------------:|--------:|------------:|------------:|-------------:|------:|----------:|----------:|--------------:|-----------:|------:|----------:|
| mlp_wass |   0.32 | 0.22 |     0.32 |    0.24 |        0.42 |    0.24 |        0.2  |        1.8  |         0.24 |  0.32 |      0.32 |      0.38 |       0.64    |       0.37 |  0.25 |      0.24 |
| mlp_js   |   0.63 | 0.34 |     0.44 |    0.8  |        0.4  |    0.33 |        0.35 |        0.51 |         0.33 |  0.67 |      0.68 |      0.65 |       0.75    |       0.62 |  0.42 |      0.67 |
| mlp_ring |   2.8  | 2.8  |     2.8  |    2.0  |        1.5  |    2.3  |        2.6  |        2.7  |         2.5  |  2.8  |      2.8  |      2.8  |       1.7e+03 |       2.8  |  2.8  |      2.4  |

| models   |   dumb |   cl |   cl_dis |   cl_v1 |   cl_v1_dis |   cl_v2 |   cl_v2_dis |   cl_v2_lin |   cl_v2_quad |   mmd |   mmd_dis |   mmd_lin |   mmd_lin_dis |   mmd_quad |   cmd |   cmd_dis |
|:---------|-------:|-----:|---------:|--------:|------------:|--------:|------------:|------------:|-------------:|------:|----------:|----------:|--------------:|-----------:|------:|----------:|
| mlp_wass |   0.32 | 0.22 |     0.31 |   0.85  |     3.4e+03 |    0.23 |        0.2  |        0.25 |         0.24 |  0.32 |      0.32 |      0.38 |       0.64    |       0.37 |  0.25 |      0.24 |
| mlp_js   |   0.63 | 0.34 |     0.45 |   8e+02 |     4.3e+03 |    0.25 |        0.35 |        0.51 |         0.33 |  0.67 |      0.68 |      0.65 |       0.75    |       0.62 |  0.42 |      0.67 |
| mlp_ring |   2.8  | 2.8  |     2.8  |   1.5   |     9.7e+03 |    2.3  |        2.6  |        2.7  |         2.5  |  2.8  |      2.8  |      2.8  |       1.7e+03 |       2.8  |  2.8  |      2.4  |

| models   |   dumb |   cl |   cl_dis |   cl_v1 |   cl_v1_dis |   cl_v2 |   cl_v2_dis |   cl_v2_lin |   cl_v2_quad |   mmd |   mmd_dis |   mmd_lin |   mmd_lin_dis |   mmd_quad |   cmd |   cmd_dis |
|:---------|-------:|-----:|---------:|--------:|------------:|--------:|------------:|------------:|-------------:|------:|----------:|----------:|--------------:|-----------:|------:|----------:|
| mlp_wass |   0.32 | 0.23 |     0.35 |    0.25 |        0.64 |    0.25 |        0.2  |        0.25 |         0.24 |  0.32 |      0.32 |      0.38 |       0.64    |       0.37 |  0.25 |      0.24 |
| mlp_js   |   0.63 | 0.32 |     0.46 |    0.41 |        0.43 |    0.48 |        0.35 |        0.51 |         0.33 |  0.67 |      0.68 |      0.65 |       0.75    |       0.62 |  0.42 |      0.67 |
| mlp_ring |   2.8  | 2.8  |     2.8  |    2.8  |        2.6  |    2.8  |        2.6  |        2.7  |         2.5  |  2.8  |      2.8  |      2.8  |       1.7e+03 |       2.8  |  2.8  |      2.4  |